Carbon chemistry and more

averdy opened the pull request:
## What changes does this PR introduce?
Common code for BGC models in pkg/gchem


## What is the current behaviour? 
BGC packages share code for carbon chemistry, air-sea fluxes, etc but each has their own files


## What is the new behaviour 
Move code to pkg/gchem. That way, updates are passed on to all BGC models


## Does this PR introduce a breaking change? 
No; each BGC package will call the gchem routines and users won't notice


## Other information:
Addresses issue #138 

## Suggested addition to `tag-index`

mjlosch committed:
merge dic/carbon_chem.F and bling/bling_carbon_chem.F

- new file gchem/gchem_carbon_chem.F with new subroutines names
  GCHEM_${OLD_SR_NAME}
- carbon chemistry variables are still in DIC/BLING_VARS.h (sub-optimal)
- results change slightly for global_oce_biogeo_bling, probably
  because of additional " _d " macros in merged gchem/gchem_carbon_chem.F

averdy committed:
Create empty PR

mjlosch committed:
continue to merge dic/carbon_chem.F and bling/bling_carbon_chem.F

- move duplicate defintion of fields in separate pkgs to GCHEM_VARS.h
- define arguments for conversion factor "permil" and
  "AtmosP"-field (only bling, sub-optimal)
- forward bling experiment global_oce_biogeo_bling breaks (from 14 to 9
  digits of agreement) mostly because of use of " _d " macro (in dic
  original but not in bling original), AD-experiment is even worse
  with 6 digits of agreement left, but gradient checks actually
  improve a little

mjlosch committed:
cleanup

mjlosch committed:
merge two versions of solveaphe into gchem_solveaphe.F

- the complication is that the parameters select* are part of the
  dic-namelist; for now they are awkwardly defined in GCHEM_VARS.h and
  set in dic/bling_readparams.F to preserve current setups, but we can
  decide to move the namelist parameters to data.gchem (TBD)

jm-c commented:
@jahn What is your thought about this ?

jahn commented:
Would be good to have the bling additions, such as allowing a varying surface pressure with the classic solver, as optional features (without the ALLOW_BLING).

Should we keep permil as a runtime parameter, and have a bcg-package-dependent default if it moves to gchem?  Might be good to rename it.

For darwin, I combined carbon_coeffs and carbon_coeffs_pressure_dep into one routine, [here](https://github.com/darwinproject/darwin3/blob/4f40d93b366e292dada7c1b8dd563fc716fd68c6/pkg/darwin/darwin_carbon_chem.F#L484), with conditionals around the depth-dependent parts.  Less code duplication and I find it even easier to read.  Also resolves the issue of missing fugf that bling put in separately.  A few small bits can be simplified too.

Yes, would be good to have all subroutines with a gchem_ prefix, though I would hope to use this in darwin too which would make it less of an issue.

There is a typo in the file name gchem_solveaphe.F, right?

Will the pH state variables live in gchem too?  For now it is easier if they don't (makes it easier to reproduce current behavior with the other bcg packages) but might be more logical to have them in gchem and have a pickup for them there.

I'll start adapting this to darwin to see what issues arise.

mjlosch committed:
correct filename: rename gchem_solveaphe.F to gchem_solvesaphe.F

mjlosch committed:
add prefix gchem_ to all routines, fix gchem_ad_diff.list

jahn commented:
I think it would make sense to use the darwin routine.  I made an effort to clean up and fix bugs when I put the pressure dependence in for the carbonate system.  It already has some of the bling additions.  One could use k=0 to trigger surface coefficients, and k>0 for pressure-corrected ones (right now 1 is for surface), so one can use the cell-center pressure also for the top layer for carbonate.

Since the select* parameters are now in gchem, would it make sense to have driver routines for the coefficients and the solver in gchem that then decide whether to call saphe or classic routines?  This decision is replicated in 3 different places in dic and bling right now.

Looking closer at how pH is handled in the different bcg packages, I think we'd have to keep it outside of gchem, at least for now.  The 3d field is handled differently in each package.

@averdy, what's the rationale for applying a pressure correction (atmosp) to surface fugf but not the other coefficients?

jahn committed:
Add missing subroutine argument for Solvesaphe

mjlosch committed:
finish (?) renaming subroutines to GCHEM_ ...

mjlosch committed:
fix some loose ends with untested code options

CPP-flags needs to be reworked, if we stick with this approach;
currently, there are flags common to DIC/BLING, that are not used in GCHEM

mjlosch committed:
merge gchem_carbon_coeffs with gchem_carbon_coeffs_presssure_dep

following jahn's example for darwin, changes results for bling even
more, not clear why

mjlosch committed:
Merge branch 'averdy/carbon' into carbon

jahn committed:
Apply pressure corrections for k=1

jahn committed:
Add missing tmpa1 for Aragonite

jahn committed:
Fix typos in CPP directives

jahn committed:
Fix calls to GCHEM_CARBON_COEFFS

mjlosch commented:
@jahn, thanks for the fixes. As usual, I am not careful enough, duh.

Differences with bling results sound plausible. I also found promoting the floating-point constants to double precision to be very important.

> I can push this to a separate branch if useful. Could also include verification experiments for missing total-pH and Munhoven solver cases.

Would you add extra verification experiments to this PR? Or to the separate branch?

jahn commented:
> Would you add extra verification experiments to this PR? Or to the separate branch?

Separate branch.  They require different CPP options, so are separate experiments.  I don't think we want to add so many experiments on the main branch.  But let me know if you think we should have them.

mjlosch commented:
OK, I assume that you are referring to `CARBONCHEM_SOLVESAPHE` and `CARBONCHEM_TOTALPHSCALE`? Maybe it *is* a good idea to have another verification experiment with these flags defined, as this code is otherwise very much untested. Maybe it can be added as a "secondary test" in one of the global_ocean experiments or tutorials?

I also think that the CPP-flags that are common to BLING and DIC (and DARWIN?) as the above should be then set in `GCHEM_OPTIONS.h` (and maybe renamed?), but this is not done (similar to runtime flags that are now defined in GCHEM_VARS.h but set in data.bling/dic), because it would break setups.

So maybe a tentative ToDo list:
- sort out differences between original bling and merged carbon chemistry (or maybe it does not matter? @averdy?)
- sort out common cpp flags (currently it's a mix with error prone inclusion of `BLING/DIC_OPTIONS.h` into `gchem`-routines)
- find a good place for common runtime parameters (in `pkg/gchem`?)
- ... (Darwin related modifications?)

mjlosch committed:
Merge upstream/master into carbon to resolve conflict

while I still remember, where this comes from

mjlosch committed:
small adjustments, avoid modifying input variables

mjlosch commented:
What is your opinion about this

> - sort out common cpp flags (currently it's a mix with error prone inclusion of BLING/DIC_OPTIONS.h into gchem-routines)

Should we move
 ```C ABIOTIC OPTIONS
C Compile "Solvesaphe" package (Munhoven 2013) for pH/pCO2
C  can still select Follows et al (2006) solver in data.bling,
C  but will use solvesaphe dissociation coefficient options
#undef CARBONCHEM_SOLVESAPHE

C In S/R CARBON_CHEM convert ak1 and ak2 to the total pH scale
C  consistent with other coefficients (currently on the seawater scale).
C NOTE: Has NO effect when CARBONCHEM_SOLVESAPHE is defined (different
C  coeffs are used).
#undef CARBONCHEM_TOTALPHSCALE
```
to `GCHEM_OPTIONS.h`? Would require to include `GCHEM_OPTIONS.h` in 4 bling-files and 4 dic-files (and probably also darwin).

> - find a good place for common runtime parameters (in pkg/gchem?)

The 5 common runtime parameters for `CARBONCHEM_SOLVESAPHE` defined
```
     selectBTconst, selectFTconst, selectHFconst, selectK1K2const, selectPHsolver
```
are now defined in `GCHEM_VARS.h` but can be set in namelists `ABIOTIC_PARMS` (same name) in `bling/dic_readparms.F`. Should we move to a common namelist in`gchem_readparms.F` (maybe `GCHEM_CARBON_PARMS`)?

I think in both cases, that moving flags and parameters to `pkg/gchem` makes sense, because that's where the carbon-pkg code in now (in this PR). The downside is clearly, that this breaks probably all setups.

jahn commented:
I agree, it is much cleaner to have these in gchem.  Just have to make sure to check for misplaced cpp options.

We could move the Munhoven parameters into a separate header and have driver routines in gchem where Munhoven vs Follows et al solver/coeffs is decided.  Then we would only have to include the header with fugf and the carbonate saturation parameters which wouldn't require GCHEM_OPTIONS.h.

seamanticscience commented:
Since `CARBONCHEM_TOTALPHSCALE` switches on/off a bug fix, would it not be better to just fix the bug in the coefficient code for `pkg/gchem` and forget about it?

seamanticscience commented:
> We could move the Munhoven parameters into a separate header and have driver routines in gchem where Munhoven vs Follows et al solver/coeffs is decided. Then we would only have to include the header with fugf and the carbonate saturation parameters which wouldn't require GCHEM_OPTIONS.h.

This sounds good for the solver routines themselves, but I think that would break the ability to use more modern carbon cycle coefficients defined in SolveSaphe with the Follows solver, wouldn't it?

mjlosch commented:
Not sure, how much of this I understand, but it appears to me (from reading the dic and bling-code), that cpp-flag `CARBONCHEM_SOLVESAPHE` is not needed at all and can be replaced with `IF (selectPHsolver.GT.0) THEN ... ELSE ... ENDIF` where this does not already happen. Except for TAF, which then requires to store all coefficient fields, basically most of the common block `CARBON_CHEM`, and field `pCO2`.

If there's discussion about the `CARBONCHEM_TOTALPHSCALE` code: the CPP-flag `CARBONCHEM_TOTALPHSCALE` is only used in `gchem_carbon_chem.F`. We have flags that we only define locally, especially when it's about locally tested code, eg. in `calc_gw.F` for `find_rho.F`.

jahn commented:
I like the idea of having the solver switch runtime.  We could have new runtime switch for the mixed case of Follows solver with modern coefficients, something like selectModernCoeffs (to be used with selectPHsolver=0).  I find that more transparent than defining CARBONCHEM_SOLVESAPHE and setting selectPHsolver=0.

jahn committed:
Choose modern carbon coefficients at runtime

and always compile Munhoven solver

jahn committed:
Update verification for runtime carbon coeff selection

jahn commented:
Done.  I've put the extra (and modified) verification experiments [here](https://github.com/jahn/verification_carbon).

I used an integer selectCarbonCoeffs.  One could also have a boolean useModernCoeffs, but then I cannot check for a user trying to select old coefficients with the Munhoven solver, I can only force useModernCoeffs to true when selectPHsolver>0.  Might still be better.  There is a branch for this in the verification repo.  Can push code changes here also.

mjlosch committed:
comment out calls to solvesaphe code ifdef ALLOW_AUTODIFF

this avoids > O(60) TAF RECOMPUTATION Warnings and ERRORs, which we
may have to sort out later

mjlosch committed:
adjust local BLING_OPTIONS.h

mjlosch commented:
Great! @jahn, do you want to add your `global_oce_biogeo_bling.saphe` (from https://github.com/jahn/verification_carbon/) yourself?

I'll leave sorting out the CARBONCHEM_TOTALPHSCALE to you @seamanticscience @jahn @averdy
We'll need new store directives with this option defined, and I'll look into that, but that's fairly independent of where you want to have this code used be default or not.

I can try to move the `select*` flags from `dic/bling_readparms.F` to `gchem_readparms.F`, if no-one objects.

mjlosch commented:
I am thinking about the AD for the case of `selectPHsolver,selectModernCoeffs>0` and in this context, I have a stupid question: `s/r gchem_carbon_chem` computes some coefficients (`ak0`,`ak1`,`ak2`,`akw`,`akb`, ...) stored in common block `CARBON_CHEM`, hence they are not passed as an argument and the loops are done within the routine. These coefficients are then passed to `s/r  GCHEM_CALC_PCO2_APPROX` (or `s/r  GCHEM_CALC_PCO2`, which both include the common block, so that they need to have different names inside the routines) for each grid point. And the loop is done outside of these routines. For `selectPHsolver,selectModernCoeffs>0`, coefficients are not passed to the corresponding routines, but used inside for individual grid points, so that the loops are again outside of the routines. Is this differential treatment/useage of loops and common blocks really necessary?

As far as I can see, we have two cases (each called three times): 
1. `selectPHsolver,selectModernCoeffs>0`: 
```
     call gchem_coeffs_surf
     [call GCHEM_COEFFS_DEEP ] sometimes
     [CALL GCHEM_AHINI_FOR_AT ] sometimes
     CALL GCHEM_CALC_PCO2_SOLVESAPHE
```
2. otherwise, the call sequence is: 
```
      call gchem_carbon_coeffs
      call gchem_calc_pco2_approx
```

Or am I getting this wrong because `selectModernCoeffs>0` does not necessarily imply `selectPHsolver>0`? 

This would be easier to read and organise, if all routines would use the same structure in terms of loop-pushing and use of common block (I guess I prefer pushing the loops inside the routines). Why do I care? From a taf-point-of-view it's also easier to deal with the first case without affecting the second (the second one does not need any additional store directives).
Further, there seems to be some duplication of code between `s/r gchem_carbon_coeffs` and `s/r gchem_coeffs_surf/deep`, but I don't dare to dive into that.

All of this is maybe something for a different PR (pure refactoring), once the fundamental issues of this PR are sorted out.

jahn committed:
Add subexp with Munhoven solver to bling verif

jahn commented:
Added it.  We could test a different set of options for the Munhoven solver from so_box_biogeo.  Feel free to change it if you have any ideas, @seamanticscience .

mjlosch committed:
add TAF store directives for case CARBONCHEM_TOTALPHSOLVE

since s/r gchem_carbon_coeffs is called multiple times, it's probably
simpler to use a local tape.

mjlosch committed:
move solvesaphe select-parameters to data.gchem namelist

selectBTconst, selectFTconst, selectHFconst, selectK1K2const,
selectPHsolver, selectModernCoeffs are defined in pkg/gchem, so it
makes sense to also set their values in gchem_readparms.F

averdy commented:
> @averdy, what's the rationale for applying a pressure correction (atmosp) to surface fugf but not the other coefficients?

@jahn If pressure correction terms are missing, we'd like to fix that. Ideally pH calculations in BLING would be the same as in Darwin, and would be in the common routines...

averdy commented:
> > > @averdy, what's the rationale for applying a pressure correction (atmosp) to surface fugf but not the other coefficients?
> > 
> > 
> > @jahn If pressure correction terms are missing, we'd like to fix that. Ideally pH calculations in BLING would be the same as in Darwin, and would be in the common routines...
> 
> As of now, darwin does not apply an atmosP correction anywhere, only water pressure.

then I suggest we drop the atmosP correction in bling

mjlosch committed:
Merge branch 'upstream/master' into carbon
